File: .agent/skills/angular-design-system-analyzer/scripts/generate_ai_rules.py
#!/usr/bin/env python3
"""
Generate AI rules for code review and component creation.
Creates guidelines based on detected patterns and conventions.
"""

import json
import sys
from pathlib import Path
from typing import Dict, List, Any


def generate_ai_rules(
    structure_file: str,
    components_file: str,
    styles_file: str,
    patterns_file: str,
    output_file: str = "ai_rules.md",
) -> str:
    """
    Generate AI rules based on analysis files.

    Args:
        structure_file: Path to structure analysis JSON
        components_file: Path to components analysis JSON
        styles_file: Path to styles analysis JSON
        patterns_file: Path to patterns analysis JSON
        output_file: Output file path

    Returns:
        Path to generated rules file
    """
    # Load analysis files
    structure = load_json(structure_file)
    components = load_json(components_file)
    styles = load_json(styles_file)
    patterns = load_json(patterns_file)

    # Generate rules document
    rules_md = generate_rules_document(structure, components, styles, patterns)

    # Write to file
    with open(output_file, "w", encoding="utf-8") as f:
        f.write(rules_md)

    return output_file


def load_json(file_path: str) -> Dict[str, Any]:
    """Load JSON file."""
    try:
        with open(file_path, "r", encoding="utf-8") as f:
            return json.load(f)
    except FileNotFoundError:
        print(f"Warning: {file_path} not found, using empty data")
        return {}
    except json.JSONDecodeError as e:
        print(f"Warning: Error parsing {file_path}: {e}")
        return {}


def generate_rules_document(
    structure: Dict[str, Any],
    components: Dict[str, Any],
    styles: Dict[str, Any],
    patterns: Dict[str, Any],
) -> str:
    """Generate the AI rules markdown document."""

    doc = """# AI Rules for Angular Design System

This document contains rules and guidelines for code review and component creation in this Angular design system.

## Table of Contents

1. [Project Architecture](#project-architecture)
2. [Naming Conventions](#naming-conventions)
3. [Component Creation Rules](#component-creation-rules)
4. [Styling Guidelines](#styling-guidelines)
5. [PrimeNG Customization Rules](#primeng-customization-rules)
6. [Code Review Checklist](#code-review-checklist)
7. [Best Practices](#best-practices)

---

## Project Architecture

"""

    # Add architecture information
    doc += generate_architecture_section(structure, patterns)

    doc += "\n---\n\n## Naming Conventions\n\n"
    doc += generate_naming_conventions_section(patterns)

    doc += "\n---\n\n## Component Creation Rules\n\n"
    doc += generate_component_rules_section(components, patterns)

    doc += "\n---\n\n## Styling Guidelines\n\n"
    doc += generate_styling_guidelines_section(styles)

    doc += "\n---\n\n## PrimeNG Customization Rules\n\n"
    doc += generate_primeng_rules_section(components, styles)

    doc += "\n---\n\n## Code Review Checklist\n\n"
    doc += generate_code_review_checklist(patterns, components, styles)

    doc += "\n---\n\n## Best Practices\n\n"
    doc += generate_best_practices_section(patterns)

    return doc


def generate_architecture_section(
    structure: Dict[str, Any], patterns: Dict[str, Any]
) -> str:
    """Generate architecture section."""
    section = ""

    folder_patterns = patterns.get("folder_structure_patterns", {})
    org_type = folder_patterns.get("organization_type", "unknown")

    section += f"### Organization Type\n\n"
    section += f"This project follows a **{org_type}** architecture.\n\n"

    if structure.get("statistics"):
        stats = structure["statistics"]
        section += f"### Project Statistics\n\n"
        section += f"- **Modules**: {stats.get('total_modules', 0)}\n"
        section += f"- **Components**: {stats.get('total_components', 0)}\n"
        section += f"- **Services**: {stats.get('total_services', 0)}\n"
        section += f"- **Directives**: {stats.get('total_directives', 0)}\n"
        section += f"- **Pipes**: {stats.get('total_pipes', 0)}\n\n"

    arch_patterns = patterns.get("architectural_patterns", {})
    if arch_patterns:
        section += f"### Architectural Patterns\n\n"
        if arch_patterns.get("uses_services"):
            section += "- ✅ Uses **Services** for business logic and data access\n"
        if arch_patterns.get("uses_state_management"):
            lib = arch_patterns.get("state_management_library", "Unknown")
            section += f"- ✅ Uses **{lib}** for state management\n"
        if arch_patterns.get("uses_lazy_loading"):
            section += "- ✅ Uses **Lazy Loading** for modules\n"
        if arch_patterns.get("uses_standalone_components"):
            section += "- ✅ Uses **Standalone Components**\n"
        if arch_patterns.get("uses_reactive_forms"):
            section += "- ✅ Uses **Reactive Forms**\n"
        section += "\n"

    return section


def generate_naming_conventions_section(patterns: Dict[str, Any]) -> str:
    """Generate naming conventions section."""
    section = ""

    naming = patterns.get("naming_conventions", {})

    section += "### File Naming\n\n"
    file_naming = naming.get("file_naming", {})
    section += f"- **Pattern**: {file_naming.get('pattern', 'kebab-case')}\n"

    common_suffixes = file_naming.get("common_suffixes", {})
    if common_suffixes:
        section += "- **Common suffixes**:\n"
        for suffix, count in list(common_suffixes.items())[:5]:
            section += f"  - `.{suffix}.ts`\n"
    section += "\n"

    section += "### Class Naming\n\n"
    class_naming = naming.get("class_naming", {})
    section += f"- **Pattern**: {class_naming.get('pattern', 'PascalCase')}\n"
    section += "- **Examples**:\n"
    section += "  - Components: `ButtonComponent`, `CardComponent`\n"
    section += "  - Services: `DataService`, `AuthService`\n"
    section += "  - Directives: `HighlightDirective`\n\n"

    section += "### Variable Naming\n\n"
    var_naming = naming.get("variable_naming", {})
    section += f"- **Pattern**: {var_naming.get('pattern', 'camelCase')}\n"
    section += f"- **Constants**: {var_naming.get('constants', 'UPPER_SNAKE_CASE')}\n"
    section += (
        f"- **Private members**: {var_naming.get('private_members', 'camelCase')}\n\n"
    )

    return section


def generate_component_rules_section(
    components: Dict[str, Any], patterns: Dict[str, Any]
) -> str:
    """Generate component creation rules."""
    section = ""

    section += "### Component Structure\n\n"
    section += "Every component MUST include:\n\n"
    section += "1. **TypeScript file** (`.component.ts`)\n"
    section += "2. **Template file** (`.component.html`)\n"
    section += "3. **Style file** (`.component.scss` or `.component.css`)\n"
    section += "4. **Test file** (`.component.spec.ts`) - if testing is configured\n\n"

    section += "### Component Metadata\n\n"
    section += "```typescript\n"
    section += "@Component({\n"
    section += "  selector: 'app-component-name',  // Use kebab-case with app- prefix\n"
    section += "  templateUrl: './component-name.component.html',\n"
    section += "  styleUrls: ['./component-name.component.scss']\n"
    section += "})\n"
    section += "export class ComponentNameComponent implements OnInit {\n"
    section += "  // Component implementation\n"
    section += "}\n"
    section += "```\n\n"

    # Add common inputs/outputs if available
    comp_patterns = components.get("component_patterns", {})
    common_inputs = comp_patterns.get("common_inputs", {})
    common_outputs = comp_patterns.get("common_outputs", {})

    if common_inputs:
        section += "### Common Inputs\n\n"
        section += "These inputs are commonly used across components:\n\n"
        for input_name in list(common_inputs.keys())[:5]:
            section += f"- `@Input() {input_name}`\n"
        section += "\n"

    if common_outputs:
        section += "### Common Outputs\n\n"
        section += "These outputs are commonly used across components:\n\n"
        for output_name in list(common_outputs.keys())[:5]:
            section += f"- `@Output() {output_name}`\n"
        section += "\n"

    # Lifecycle hooks
    common_hooks = comp_patterns.get("common_lifecycle_hooks", {})
    if common_hooks:
        section += "### Common Lifecycle Hooks\n\n"
        for hook in list(common_hooks.keys())[:5]:
            section += f"- `{hook}()`\n"
        section += "\n"

    return section


def generate_styling_guidelines_section(styles: Dict[str, Any]) -> str:
    """Generate styling guidelines."""
    section = ""

    stats = styles.get("statistics", {})

    section += "### Style File Organization\n\n"
    section += f"- **Global styles**: {stats.get('global_styles', 0)} files\n"
    section += f"- **Component styles**: {stats.get('component_styles', 0)} files\n"
    section += f"- **Theme files**: {stats.get('theme_files', 0)} files\n\n"

    section += "### Variables\n\n"
    section += f"- **SCSS variables**: {stats.get('scss_variables', 0)} defined\n"
    section += (
        f"- **CSS custom properties**: {stats.get('css_variables', 0)} defined\n\n"
    )

    # Color palette
    color_palette = styles.get("color_palette", {})
    if color_palette:
        section += "### Color Palette\n\n"
        section += "Use these predefined color variables:\n\n"
        for var_name, var_value in list(color_palette.items())[:10]:
            section += f"- `{var_name}`: {var_value}\n"
        if len(color_palette) > 10:
            section += f"\n... and {len(color_palette) - 10} more color variables\n"
        section += "\n"

    # Mixins
    mixins = styles.get("mixins", [])
    if mixins:
        section += "### Available Mixins\n\n"
        unique_mixins = {m["name"]: m for m in mixins}.values()
        for mixin in list(unique_mixins)[:5]:
            section += f"- `@mixin {mixin['name']}{mixin.get('parameters', '')}`\n"
        section += "\n"

    section += "### Styling Rules\n\n"
    section += "1. **Always use SCSS variables** for colors, spacing, and typography\n"
    section += "2. **Avoid hardcoded values** - use design tokens\n"
    section += "3. **Component styles should be scoped** to the component\n"
    section += "4. **Use mixins** for reusable style patterns\n"
    section += "5. **Follow BEM naming** for CSS classes when appropriate\n\n"

    return section


def generate_primeng_rules_section(
    components: Dict[str, Any], styles: Dict[str, Any]
) -> str:
    """Generate PrimeNG customization rules."""
    section = ""

    primeng_comps = components.get("primeng_components_used", [])

    if primeng_comps:
        section += "### PrimeNG Components Used\n\n"
        section += "This design system uses the following PrimeNG components:\n\n"
        for comp in primeng_comps[:15]:
            section += f"- `{comp}`\n"
        if len(primeng_comps) > 15:
            section += f"\n... and {len(primeng_comps) - 15} more\n"
        section += "\n"

    section += "### Customization Guidelines\n\n"
    section += "1. **Override PrimeNG styles** in dedicated theme files\n"
    section += (
        "2. **Use `::ng-deep`** carefully - prefer component-level customization\n"
    )
    section += (
        "3. **Create wrapper components** for heavily customized PrimeNG components\n"
    )
    section += "4. **Document all customizations** in component comments\n"
    section += "5. **Test customizations** across all PrimeNG themes\n\n"

    # PrimeNG overrides
    primeng_overrides = styles.get("primeng_overrides", [])
    if primeng_overrides:
        section += "### Common PrimeNG Overrides\n\n"
        section += (
            f"Found {len(primeng_overrides)} files with PrimeNG style overrides.\n\n"
        )
        section += "**Example override patterns**:\n\n"
        section += "```scss\n"
        section += "// Override PrimeNG button styles\n"
        section += "::ng-deep .p-button {\n"
        section += "  border-radius: $border-radius;\n"
        section += "  padding: $button-padding;\n"
        section += "}\n"
        section += "```\n\n"

    return section


def generate_code_review_checklist(
    patterns: Dict[str, Any], components: Dict[str, Any], styles: Dict[str, Any]
) -> str:
    """Generate code review checklist."""
    section = ""

    section += "Use this checklist when reviewing code:\n\n"

    section += "### General\n\n"
    section += "- [ ] Code follows naming conventions\n"
    section += "- [ ] No console.log statements in production code\n"
    section += "- [ ] Proper error handling implemented\n"
    section += "- [ ] Code is properly formatted and linted\n\n"

    section += "### Components\n\n"
    section += "- [ ] Component selector follows naming convention (kebab-case)\n"
    section += "- [ ] Component has proper lifecycle hooks\n"
    section += "- [ ] Inputs and outputs are properly typed\n"
    section += "- [ ] Component is properly documented\n"
    section += "- [ ] Template and styles are in separate files\n"
    section += "- [ ] OnPush change detection used when appropriate\n\n"

    section += "### Styles\n\n"
    section += "- [ ] Uses SCSS variables instead of hardcoded values\n"
    section += "- [ ] No !important unless absolutely necessary\n"
    section += "- [ ] Follows established color palette\n"
    section += "- [ ] Responsive design considerations\n"
    section += "- [ ] PrimeNG overrides are properly scoped\n\n"

    section += "### Services\n\n"
    section += "- [ ] Service is properly injectable\n"
    section += "- [ ] Proper dependency injection\n"
    section += "- [ ] Error handling for HTTP calls\n"
    section += "- [ ] Observables properly managed (unsubscribe)\n\n"

    section += "### Testing\n\n"
    section += "- [ ] Unit tests exist and pass\n"
    section += "- [ ] Test coverage is adequate\n"
    section += "- [ ] Edge cases are tested\n\n"

    return section


def generate_best_practices_section(patterns: Dict[str, Any]) -> str:
    """Generate best practices section."""
    section = ""

    section += "### TypeScript\n\n"
    section += "- Use **strict type checking**\n"
    section += "- Prefer **interfaces** over type aliases for object shapes\n"
    section += "- Use **enums** for fixed sets of values\n"
    section += "- Avoid **any** type - use **unknown** if type is truly unknown\n\n"

    section += "### Angular Specific\n\n"
    section += "- Use **OnPush change detection** for performance\n"
    section += "- Implement **trackBy** functions for ngFor loops\n"
    section += "- Use **async pipe** for observables in templates\n"
    section += "- Unsubscribe from observables to prevent memory leaks\n"
    section += "- Use **lazy loading** for feature modules\n\n"

    section += "### RxJS\n\n"
    section += "- Prefer **declarative** approach over imperative\n"
    section += "- Use **operators** like map, filter, switchMap appropriately\n"
    section += "- Handle errors with **catchError** operator\n"
    section += "- Use **shareReplay** to avoid multiple HTTP calls\n\n"

    section += "### Performance\n\n"
    section += "- Minimize **bundle size** - use lazy loading\n"
    section += "- Optimize **change detection** strategy\n"
    section += "- Use **pure pipes** when possible\n"
    section += "- Avoid complex logic in templates\n"
    section += "- Use **virtual scrolling** for long lists\n\n"

    return section


def main():
    if len(sys.argv) < 5:
        print(
            "Usage: python generate_ai_rules.py <structure.json> <components.json> <styles.json> <patterns.json> [output.md]"
        )
        print(
            "Example: python generate_ai_rules.py structure.json components.json styles.json patterns.json ai_rules.md"
        )
        sys.exit(1)

    structure_file = sys.argv[1]
    components_file = sys.argv[2]
    styles_file = sys.argv[3]
    patterns_file = sys.argv[4]
    output_file = sys.argv[5] if len(sys.argv) > 5 else "ai_rules.md"

    try:
        print(f"Generating AI rules from analysis files...")
        output_path = generate_ai_rules(
            structure_file, components_file, styles_file, patterns_file, output_file
        )

        print(f"\n✅ AI rules generated successfully!")
        print(f"   Output: {output_path}")

    except Exception as e:
        print(f"❌ Error: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: .agent/skills/angular-design-system-analyzer-v2/scripts/generate_ai_rules.mjs
#!/usr/bin/env node
/**
 * Generate AI rules for code review and component creation.
 * Creates guidelines based on detected patterns and conventions.
 */
import fs from 'node:fs';

function loadJson(filePath) {
  try { return JSON.parse(fs.readFileSync(filePath, 'utf-8')); }
  catch (e) {
    if (e.code === 'ENOENT') { console.warn(`Warning: ${filePath} not found, using empty data`); return {}; }
    console.warn(`Warning: Error parsing ${filePath}: ${e.message}`); return {};
  }
}

function generateAiRules(structureFile, componentsFile, stylesFile, patternsFile, outputFile = 'ai_rules.md') {
  const structure = loadJson(structureFile);
  const components = loadJson(componentsFile);
  const styles = loadJson(stylesFile);
  const patterns = loadJson(patternsFile);
  const rulesMd = generateRulesDocument(structure, components, styles, patterns);
  fs.writeFileSync(outputFile, rulesMd, 'utf-8');
  return outputFile;
}

function generateRulesDocument(structure, components, styles, patterns) {
  let doc = `# AI Rules for Angular Design System

This document contains rules and guidelines for code review and component creation in this Angular design system.

## Table of Contents

1. [Project Architecture](#project-architecture)
2. [Naming Conventions](#naming-conventions)
3. [Component Creation Rules](#component-creation-rules)
4. [Styling Guidelines](#styling-guidelines)
5. [PrimeNG Customization Rules](#primeng-customization-rules)
6. [Code Review Checklist](#code-review-checklist)
7. [Best Practices](#best-practices)

---

## Project Architecture

`;
  doc += genArchSection(structure, patterns);
  doc += '\n---\n\n## Naming Conventions\n\n';
  doc += genNamingSection(patterns);
  doc += '\n---\n\n## Component Creation Rules\n\n';
  doc += genComponentSection(components, patterns);
  doc += '\n---\n\n## Styling Guidelines\n\n';
  doc += genStylingSection(styles);
  doc += '\n---\n\n## PrimeNG Customization Rules\n\n';
  doc += genPrimengSection(components, styles);
  doc += '\n---\n\n## Code Review Checklist\n\n';
  doc += genChecklist(patterns, components, styles);
  doc += '\n---\n\n## Best Practices\n\n';
  doc += genBestPractices(patterns);
  return doc;
}

function genArchSection(structure, patterns) {
  let s = '';
  const fp = patterns.folder_structure_patterns || {};
  s += `### Organization Type\n\nThis project follows a **${fp.organization_type || 'unknown'}** architecture.\n\n`;
  const stats = structure.statistics;
  if (stats) {
    s += '### Project Statistics\n\n';
    s += `- **Modules**: ${stats.total_modules || 0}\n`;
    s += `- **Components**: ${stats.total_components || 0}\n`;
    s += `- **Services**: ${stats.total_services || 0}\n`;
    s += `- **Directives**: ${stats.total_directives || 0}\n`;
    s += `- **Pipes**: ${stats.total_pipes || 0}\n\n`;
  }
  const ap = patterns.architectural_patterns || {};
  if (Object.keys(ap).length) {
    s += '### Architectural Patterns\n\n';
    if (ap.uses_services) s += '- ✅ Uses **Services** for business logic and data access\n';
    if (ap.uses_state_management) s += `- ✅ Uses **${ap.state_management_library || 'Unknown'}** for state management\n`;
    if (ap.uses_lazy_loading) s += '- ✅ Uses **Lazy Loading** for modules\n';
    if (ap.uses_standalone_components) s += '- ✅ Uses **Standalone Components**\n';
    if (ap.uses_reactive_forms) s += '- ✅ Uses **Reactive Forms**\n';
    s += '\n';
  }
  return s;
}

function genNamingSection(patterns) {
  let s = '';
  const n = patterns.naming_conventions || {};
  const fn = n.file_naming || {};
  s += '### File Naming\n\n';
  s += `- **Pattern**: ${fn.pattern || 'kebab-case'}\n`;
  const cs = fn.common_suffixes || {};
  if (Object.keys(cs).length) { s += '- **Common suffixes**:\n'; for (const sf of Object.keys(cs).slice(0, 5)) s += `  - \`.${sf}.ts\`\n`; }
  s += '\n### Class Naming\n\n';
  const cn = n.class_naming || {};
  s += `- **Pattern**: ${cn.pattern || 'PascalCase'}\n`;
  s += '- **Examples**:\n  - Components: `ButtonComponent`, `CardComponent`\n  - Services: `DataService`, `AuthService`\n  - Directives: `HighlightDirective`\n\n';
  s += '### Variable Naming\n\n';
  const vn = n.variable_naming || {};
  s += `- **Pattern**: ${vn.pattern || 'camelCase'}\n`;
  s += `- **Constants**: ${vn.constants || 'UPPER_SNAKE_CASE'}\n`;
  s += `- **Private members**: ${vn.private_members || 'camelCase'}\n\n`;
  return s;
}

function genComponentSection(components, patterns) {
  let s = '### Component Structure\n\nEvery component MUST include:\n\n';
  s += '1. **TypeScript file** (`.component.ts`)\n';
  s += '2. **Template file** (`.component.html`)\n';
  s += '3. **Style file** (`.component.scss` or `.component.css`)\n';
  s += '4. **Test file** (`.component.spec.ts`) - if testing is configured\n\n';
  s += `### Component Metadata

\`\`\`typescript
@Component({
  selector: 'app-component-name',
  templateUrl: './component-name.component.html',
  styleUrls: ['./component-name.component.scss']
})
export class ComponentNameComponent implements OnInit {
  // Component implementation
}
\`\`\`

`;
  const cp = components.component_patterns || {};
  const ci = cp.common_inputs || {};
  if (Object.keys(ci).length) { s += '### Common Inputs\n\nThese inputs are commonly used across components:\n\n'; for (const i of Object.keys(ci).slice(0, 5)) s += `- \`@Input() ${i}\`\n`; s += '\n'; }
  const co = cp.common_outputs || {};
  if (Object.keys(co).length) { s += '### Common Outputs\n\nThese outputs are commonly used across components:\n\n'; for (const o of Object.keys(co).slice(0, 5)) s += `- \`@Output() ${o}\`\n`; s += '\n'; }
  const ch = cp.common_lifecycle_hooks || {};
  if (Object.keys(ch).length) { s += '### Common Lifecycle Hooks\n\n'; for (const h of Object.keys(ch).slice(0, 5)) s += `- \`${h}()\`\n`; s += '\n'; }
  return s;
}

function genStylingSection(styles) {
  let s = '';
  const st = styles.statistics || {};
  s += '### Style File Organization\n\n';
  s += `- **Global styles**: ${st.global_styles || 0} files\n`;
  s += `- **Component styles**: ${st.component_styles || 0} files\n`;
  s += `- **Theme files**: ${st.theme_files || 0} files\n\n`;
  s += '### Variables\n\n';
  s += `- **SCSS variables**: ${st.scss_variables || 0} defined\n`;
  s += `- **CSS custom properties**: ${st.css_variables || 0} defined\n\n`;
  const cp = styles.color_palette || {};
  const entries = Object.entries(cp);
  if (entries.length) {
    s += '### Color Palette\n\nUse these predefined color variables:\n\n';
    for (const [vn, vv] of entries.slice(0, 10)) s += `- \`${vn}\`: ${vv}\n`;
    if (entries.length > 10) s += `\n... and ${entries.length - 10} more color variables\n`;
    s += '\n';
  }
  const mx = styles.mixins || [];
  if (mx.length) {
    const unique = Object.values(mx.reduce((a, m) => { a[m.name] = m; return a; }, {}));
    s += '### Available Mixins\n\n';
    for (const m of unique.slice(0, 5)) s += `- \`@mixin ${m.name}${m.parameters || ''}\`\n`;
    s += '\n';
  }
  s += `### Styling Rules

1. **Always use SCSS variables** for colors, spacing, and typography
2. **Avoid hardcoded values** - use design tokens
3. **Component styles should be scoped** to the component
4. **Use mixins** for reusable style patterns
5. **Follow BEM naming** for CSS classes when appropriate

`;
  return s;
}

function genPrimengSection(components, styles) {
  let s = '';
  const pc = components.primeng_components_used || [];
  if (pc.length) {
    s += '### PrimeNG Components Used\n\nThis design system uses the following PrimeNG components:\n\n';
    for (const c of pc.slice(0, 15)) s += `- \`${c}\`\n`;
    if (pc.length > 15) s += `\n... and ${pc.length - 15} more\n`;
    s += '\n';
  }
  s += `### Customization Guidelines

1. **Override PrimeNG styles** in dedicated theme files
2. **Use \`::ng-deep\`** carefully - prefer component-level customization
3. **Create wrapper components** for heavily customized PrimeNG components
4. **Document all customizations** in component comments
5. **Test customizations** across all PrimeNG themes

`;
  const po = styles.primeng_overrides || [];
  if (po.length) {
    s += `### Common PrimeNG Overrides

Found ${po.length} files with PrimeNG style overrides.

**Example override patterns**:

\`\`\`scss
// Override PrimeNG button styles
::ng-deep .p-button {
  border-radius: $border-radius;
  padding: $button-padding;
}
\`\`\`

`;
  }
  return s;
}

function genChecklist() {
  return `Use this checklist when reviewing code:

### General

- [ ] Code follows naming conventions
- [ ] No console.log statements in production code
- [ ] Proper error handling implemented
- [ ] Code is properly formatted and linted

### Components

- [ ] Component selector follows naming convention (kebab-case)
- [ ] Component has proper lifecycle hooks
- [ ] Inputs and outputs are properly typed
- [ ] Component is properly documented
- [ ] Template and styles are in separate files
- [ ] OnPush change detection used when appropriate

### Styles

- [ ] Uses SCSS variables instead of hardcoded values
- [ ] No !important unless absolutely necessary
- [ ] Follows established color palette
- [ ] Responsive design considerations
- [ ] PrimeNG overrides are properly scoped

### Services

- [ ] Service is properly injectable
- [ ] Proper dependency injection
- [ ] Error handling for HTTP calls
- [ ] Observables properly managed (unsubscribe)

### Testing

- [ ] Unit tests exist and pass
- [ ] Test coverage is adequate
- [ ] Edge cases are tested

`;
}

function genBestPractices() {
  return `### TypeScript

- Use **strict type checking**
- Prefer **interfaces** over type aliases for object shapes
- Use **enums** for fixed sets of values
- Avoid **any** type - use **unknown** if type is truly unknown

### Angular Specific

- Use **OnPush change detection** for performance
- Implement **trackBy** functions for ngFor loops
- Use **async pipe** for observables in templates
- Unsubscribe from observables to prevent memory leaks
- Use **lazy loading** for feature modules

### RxJS

- Prefer **declarative** approach over imperative
- Use **operators** like map, filter, switchMap appropriately
- Handle errors with **catchError** operator
- Use **shareReplay** to avoid multiple HTTP calls

### Performance

- Minimize **bundle size** - use lazy loading
- Optimize **change detection** strategy
- Use **pure pipes** when possible
- Avoid complex logic in templates
- Use **virtual scrolling** for long lists

`;
}

// --- Main ---
const args = process.argv.slice(2);
if (args.length < 4) {
  console.log('Usage: node generate_ai_rules.mjs <structure.json> <components.json> <styles.json> <patterns.json> [output.md]');
  process.exit(1);
}
try {
  console.log('Generating AI rules from analysis files...');
  const out = generateAiRules(args[0], args[1], args[2], args[3], args[4] || 'ai_rules.md');
  console.log(`\n✅ AI rules generated successfully!\n   Output: ${out}`);
} catch (e) { console.error(`❌ Error: ${e.message}`); process.exit(1); }
